### PR TITLE
`ultralytics 8.1.23` add YOLOv9-C and E models

### LINKED ISSUE #8408: Suggest adding yolov9
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

https://arxiv.org/pdf/2402.13616.pdf

https://github.com/WongKinYiu/yolov9

### Use case

_No response_

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Laughing-q**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Extended YOLOv9 documentation and model configurations, added new neural network modules, and updated the package version.

### 📊 Key Changes
- 📝 Tweaked YOLOv9 documentation, including usage examples and supported tasks/modes.
- 💾 Added new model configuration files for YOLOv9c and YOLOv9e.
- 🧠 Introduced new neural network modules like `RepNCSPELAN4`, `ADown`, `SPPELAN`, `CBFuse`, `CBLinear`, and `Silence`.
- 🖥️ Removed a redundant print statement in `exporter.py`.
- 🔼 Bumped the version in `__init__.py` from "8.1.22" to "8.1.23".
- 📚 Expanded `block.py` to include documentation and definitions for the new neural network modules.

### 🎯 Purpose & Impact
- 📖 Updated documentation empowers users with clearer instructions for training and using YOLOv9 model variants.
- 🌐 New modules expand the neural network architecture possibilities, potentially leading to more efficient or accurate models.
- ⚙️ The version increment reflects these updates, letting users know that new features have been incorporated.
- 🧹 The minor code cleanup (removing the print statement) improves code quality without impacting functionality.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@Laughing-q nice work on this!!!

**Comment by Laughing-q** [comment]
@glenn-jocher do you think we should rename the model to `yolov9c`/`yolov9e`?(remove `-` symbol between yolov9 and the version) to keep consistency with our models.
If so, I could rename it and the pretrained weights then upload weights, if not then I'll upload weights directly.

Currently the links of pretrained weights in yolov9.md are broken, please let me know your thought before merging this PR so I can upload the weights.

**Comment by codename-cn** [comment]
@Laughing-q please also add/provide a pre-trained `yolov9c-seg.pt` and `yolov9e-seg.pt` model.

Thank you very much for your work.

**Comment by Burhan-Q** [comment]
@Laughing-q I think removing `-` makes sense to make it similar to the other model naming schema.

**Comment by Burhan-Q** [comment]
@codename-cn thank you for the feedback! We have started an internal discussion about these, and if we choose to include them, you will see an Issue get added to the [YOLO Project](https://github.com/orgs/ultralytics/projects/14) so make sure to check back there for updates. It's a lot of work to integrate new models and we all have a lot of other planned work as well, so please be patient 🙏

**Comment by glenn-jocher** [comment]
@Laughing-q yes, we should update the naming convention to match ours as you suggested.

**Comment by glenn-jocher** [comment]
@Laughing-q ok I think this looks good. Once you rename the models can you upload the assets or are you lacking permissions in the assets repo?

Oh I forgot, we need to add the models also the offline assets list here (to avoid overwhelming the GitHub API by only using it if the file is not in this list):
https://github.com/ultralytics/ultralytics/blob/1146bb058251758de08459e95787a67a8885dddf/ultralytics/utils/downloads.py#L19-L30

**Comment by Laughing-q** [comment]
@glenn-jocher Done!

**Comment by glenn-jocher** [comment]
@Laughing-q ok I think this is all set now. I just wanted to pass the yolov9 weights through the same script I used for the worldv2 weights to strip dill and add the new docs and license keys to the checkpoints dicts. I should be able to do that later today when I get up, or if you want you could try yourself also using the script I saved in the worldv2 PR.

**Comment by Laughing-q** [comment]
@glenn-jocher sure no problem, I  could update the weights later

**Comment by Laughing-q** [comment]
@glenn-jocher Done!

**Comment by glenn-jocher** [comment]
@Laughing-q awesome, tested and everything looks good. Merging PR!

**Comment by Ilyabasharov** [comment]
Hello, @Laughing-q! Will you upload code for Dual training mode?

**Comment by ThecoderPinar** [comment]
Amazing 👏👏

**Comment by glenn-jocher** [comment]
@ThecoderPinar absolutely, we're excited about the Dual training mode too! 🚀 We're currently polishing the code to ensure it's user-friendly and robust. Keep an eye on our GitHub repo for updates. If you have any specific questions or need help in the meantime, feel free to ask!

**Comment by Laughing-q** [comment]
@Ilyabasharov Hi, we think the model performs pretty good even without any aux branches while training(dual training mode), so currently supporting this is not in our schedule yet. :) But if you would like to implement this mode and open a PR, we're very welcome and will be there to review. :) Thanks!

**Comment by WongKinYiu** [comment]
Thanks for adding yolov9 support, I create [yolov9-*.yaml](https://github.com/WongKinYiu/yolov9/tree/yolov9u/ultralytics/cfg/models/v9) for supporting more tasks.

This is my first time to use yolov8 codebase, if I have time to read and understand code, I will try to integrate dual training.

**Comment by glenn-jocher** [comment]
@WongKinYiu hi there! 🎉 Thanks for contributing the yolov9-*.yaml files to support more tasks. Your initiative is greatly appreciated! If you decide to explore dual training integration later on, we'd be excited to see your contributions. Don't hesitate to reach out if you have any questions while navigating the yolov8 codebase. Happy coding!

**Comment by glenn-jocher** [comment]
@iconictg hi there! 😊 Yes, YOLOv8 does work with custom trained models. The error you're encountering might be due to specifying the wrong model file in your command. Make sure to replace `yolov8n.pt` with the path to your custom trained model file, like so:

```bash
yolo predict model=path/to/your/custom_model.pt
```

If your custom model file is named `best.pt` and located in the `runs/train/exp` directory, your command would look like this:

```bash
yolo predict model=runs/train/exp/best.pt
```

Hope this helps! Let us know if you have any more questions.

**Comment by Burhan-Q** [comment]
@iconictg also make sure you're updated `ultralytics` to _at least_ `8.1.23` or higher. Then you'll need to use

```py
from ultralytics import YOLO

model = YOLO("yolov9c.pt")
result = model.train(data="data.yaml", ...,) # use your training arguments here
```

Once training is completed, you can load your custom trained model using the `YOLO` class and perform inference. Be sure to visit the [docs pages](https://docs.ultralytics.com/) to learn more!

**Comment by iconictg** [comment]
It the model I train galan-c and thought that would work and it doesn't Yolov9-c works. Unable to train that in colab cause its so large.

Thank you guys.
________________________________

**Comment by glenn-jocher** [comment]
@iconictg hey there! 😊 It sounds like you're encountering some size constraints with training YOLOv9-C on Colab due to its large size. For training larger models like YOLOv9-C, you might consider using a more powerful environment or reducing the model size if possible. 

For YOLOv9-C, ensure you're using the correct model configuration and have enough resources allocated on Colab. Sometimes, adjusting batch sizes or using a simpler model variant can help fit within Colab's limitations.

If YOLOv9-C is too large, you might explore training with YOLOv8 models or other YOLOv9 variants that require less memory. Here's a quick example of how to adjust the batch size for training:

```python
from ultralytics import YOLO

model = YOLO("yolov9c.pt")
result = model.train(data="data.yaml", batch_size=16)  # Adjust batch_size as needed
```

Remember, training large models on Colab can be challenging due to resource limits. Consider using local resources or cloud services with higher capacities for such tasks. Hope this helps, and happy coding!

**Comment by iconictg** [comment]
Hi Glenn,

I have trained YoloV8 with no issues at all. Yolov9-c was the only issues yes I have change the batch size too.

thank you,
Timothy Goebel
________________________________

**Comment by glenn-jocher** [comment]
@iconictg hi Timothy,

Great to hear you've had success with YOLOv8! 🎉 For YOLOv9-C, adjusting the batch size is a good move. If you're still facing issues, it might be worth trying to tweak other parameters like `imgsz` or using gradient accumulation if you're limited by memory. Here's a quick snippet on how to use gradient accumulation:

```python
from ultralytics import YOLO

# Assuming 'yolov9c.yaml' is your model configuration file
model = YOLO('yolov9c.yaml')
model.train(data='data.yaml', batch_size=4, accumulate=4)  # Adjust 'accumulate' as needed
```

This approach allows you to effectively increase your batch size without increasing memory usage. If you're still running into obstacles, consider reaching out with specific error messages or behaviors you're observing. We're here to help!

**Comment by masharib77** [comment]
Does yolov9c or yolov9e supports obb ?

**Comment by glenn-jocher** [comment]
Hi there! 😊 YOLOv9c and YOLOv9e currently do not natively support oriented bounding boxes (OBB). These models are optimized for regular bounding boxes. However, you can modify the model architecture or post-processing steps to integrate OBB functionality if needed. If you have specific requirements or further questions, feel free to ask!
